pkg/receive/handler.go

```diff
@@ -0,0 +1,204 @@
+package receive
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	stdlog "log"
+	"net"
+	"net/http"
+	"sync/atomic"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/improbable-eng/thanos/pkg/store/prompb"
+	conntrack "github.com/mwitkow/go-conntrack"
+	"github.com/opentracing-contrib/go-stdlib/nethttp"
+	opentracing "github.com/opentracing/opentracing-go"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/prometheus/common/route"
+	promtsdb "github.com/prometheus/prometheus/storage/tsdb"
+	"github.com/prometheus/tsdb"
+)
+
+var (
+	requestDuration = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "thanos_http_request_duration_seconds",
+			Help:    "Histogram of latencies for HTTP requests.",
+			Buckets: []float64{.1, .2, .4, 1, 3, 8, 20, 60, 120},
+		},
+		[]string{"handler"},
+	)
+	responseSize = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "thanos_http_response_size_bytes",
+			Help:    "Histogram of response size for HTTP requests.",
+			Buckets: prometheus.ExponentialBuckets(100, 10, 8),
+		},
+		[]string{"handler"},
+	)
+)
+
+// Options for the web Handler.
+type Options struct {
+	Receiver      *Writer
+	ListenAddress string
+	Registry      prometheus.Registerer
+	ReadyStorage  *promtsdb.ReadyStorage
+}
+
+// Handler serves a Prometheus remote write receiving HTTP endpoint.
+type Handler struct {
+	readyStorage *promtsdb.ReadyStorage
+	logger       log.Logger
+	receiver     *Writer
+	router       *route.Router
+	options      *Options
+	quitCh       chan struct{}
+
+	ready uint32 // ready is uint32 rather than boolean to be able to use atomic functions.
+}
+
+func instrumentHandler(handlerName string, handler http.HandlerFunc) http.HandlerFunc {
+	return promhttp.InstrumentHandlerDuration(
+		requestDuration.MustCurryWith(prometheus.Labels{"handler": handlerName}),
+		promhttp.InstrumentHandlerResponseSize(
+			responseSize.MustCurryWith(prometheus.Labels{"handler": handlerName}),
+			handler,
+		),
+	)
+}
+
+func NewHandler(logger log.Logger, o *Options) *Handler {
+	router := route.New().WithInstrumentation(instrumentHandler)
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+
+	h := &Handler{
+		logger:       logger,
+		router:       router,
+		readyStorage: o.ReadyStorage,
+		receiver:     o.Receiver,
+		options:      o,
+		quitCh:       make(chan struct{}),
+	}
+
+	readyf := h.testReady
+	router.Post("/receive", readyf(h.receive))
```

Comment:
Ok, so this starts to be some API. We already keep Prometheus like API to guide people and be intuitive. This means that remote receiver would be nice to use the same or similar... does Prometheus serves remote write endpoint? What endpoint looks like?  `api/v1/write` ? As a basics something like `v1` in path is useful.

Reply:
tend to agree but flexible on the url ... `api/v1/receive` feels more correct to me

Reply:
agreed with the prefix, going with `/api/v1/receive`